### DISCUSSION TITLE
Generate version file on build

### DISCUSSION THREAD

**Discussion by sebthom** [pdm-project/pdm#870] [Q&A]
Hi,
I am new to pdm. I am trying to convert a distutils/setup.py based project to pdm.

I would like to be able to specify the version (e.g. `1.0`) manually in the pyproject.toml file and when I build the project using `pdm build` a `version.py` should be generated/packaged that also includes the commit hash, similar to what `setuptools_scm` is doing, e.g.

```python
# coding: utf-8
# file generated by setuptools_scm
# don't change, don't track in version control
version = '1.0.dev12+g6961bab'
version_tuple = (1, 0, 'dev12', 'g6961bab')
```

During application runtime I want to read the version.py file and be able to display/log/evaluate the version.

How can I achieve this with pdm?

**Comment by frostming** [comment]
Dynamic versions: https://pdm.fming.dev/latest/pyproject/pep621/#determine-the-package-version-dynamically

You can inspect the version in Python via:

```python
import importlib.metadata
version = importlib.metadata.version(package_name)
```

**Reply by sebthom** [reply]
@frostming  When I use this config:
```yaml
[project]
dynamic = ["version"]

[tool.pdm]
version = {use_scm = true}
```
and build on main branch, PDM generates the version number `0.1.dev12+g6961bab.d20220127.`

Where does the `0.1` come from? How can I manually specify it but have PDM add/generate the classifier `.dev12+g6961bab.d20220127` ?

**Reply by frostming** [reply]
`use_scm` does the same as `setuptools_scm`, which takes the version from the last **SCM tag** and if not found, defaults to `0.1`. You should make sure the source is not a shallow copy.

**Reply by sebthom** [reply]
Is the `0.1` fallback value configurable? where in the code is it set?

**Reply by cmarqu** [reply]
These days, the fallback value seems to be "0.0.0" and it is coming from https://github.com/pdm-project/pdm-backend/blob/main/src/pdm/backend/hooks/version/scm.py

**Comment by berislavlopac** [comment]
It would be helpful to have the `pdm` documentation link to the `pdm-backend` docs on this topic.

**Comment by lwasser** [comment]
hey there @frostming - that link above returns a really odd redirect issue that never resolves - if there a new working link?
The link to [dynamic versioning here](https://pdm.fming.dev/latest/reference/pep621/#package-version), goes to build page with no info on dynamic versioning. 

<img width="1743" alt="Screen Shot 2023-08-17 at 10 20 22 AM" src="https://github.com/pdm-project/pdm/assets/7649194/a0b132a7-ed97-4b62-a909-d78e5bbf0911">

Many thanks,

**Reply by frostming** [reply]
Dynamic versioning is a feature of the backends rather than PDM. This part of docs needs refactoring at some point.

You should look at the documentation of `pdm-backend`, `hatchling`, `setuptools-scm` for example.

**Reply by lwasser** [reply]
thanks!! ok i found the [docs here for pdm-backend!](https://pdm-backend.fming.dev/metadata/#dynamic-project-version) in case anyone else happens upon this.
